canonical/vanilla-framework
# Update core icons to support new theming system

## Done

Updates the icons to use new CSS variables for theming.
Involved a bit of refactoring, so that both previous way of including icon mixins works and the new themed icons.

So currently icons have 2 mixins each:

`@vf-icon-NAME($color)` - that is deprecated and backwards compatible way of including an icon with an option to change its colour.

`@vf-icon-NAME-themed` - that is a new recommended way to include icon, that will use the icon colour defined by current theme (via CSS variables).

This PR also updates icon colours, to use default text colour in each theme, so black in light themes, and white in dark theme.

We also introduce `--vf-is-light-theme` and `--vf-is-dark-theme` CSS variables that are used by icon mixins to conditionally switch to correct version. This is based on a "var space hack" solution explained in this blog post: https://lea.verou.me/blog/2020/10/the-var-space-hack-to-toggle-multiple-values-with-one-custom-property/

**NOTE** New theming is only applied to standard set of Vanilla icons that are directly included in Vanilla. It doesn't cover all individual icons in the "additional" set.

Fixes [WD-8624](https://warthogs.atlassian.net/browse/WD-8624)

## QA

- Open [demo](https://vanilla-framework-4993.demos.haus/docs/examples/patterns/icons/icons-dark)
- Check if icons display correctly in dark variant
- Change the body class name to `is-light` or `is-paper`, check if icons adjust accordingly
- Review icons in context of other components to make sure they render correctly:
    - Notifications (icons and dismiss button): [Variants light](https://vanilla-framework-4993.demos.haus/docs/examples/patterns/notifications/variants) [Variants dar](https://vanilla-framework-4993.demos.haus/docs/examples/patterns/notifications/variants-dark)
    - Chips (dismiss button): [Variants](https://vanilla-framework-4993.demos.haus/docs/examples/patterns/chip/variants)
    - Side navigation (accordion chevron): [Dark](https://vanilla-framework-4993.demos.haus/docs/examples/patterns/side-navigation/accordion-dark) [Light](https://vanilla-framework-4993.demos.haus/docs/examples/patterns/side-navigation/accordion)
    - Buttons [with icons](https://vanilla-framework-4993.demos.haus/docs/examples/patterns/buttons/icon)
    - Forms validation: [light](https://vanilla-framework-4993.demos.haus/docs/examples/patterns/forms/form-validation) [dark](https://vanilla-framework-4993.demos.haus/docs/examples/patterns/forms/form-validation-dark)
- Review updated documentation:
  - https://vanilla-framework-4993.demos.haus/docs/patterns/icons
  - Mostly focus on "Dark theme" and "Customisation" sections, but make sure the rest doesn't contain any outdated information as well

### Check if PR is ready for release

If this PR contains Vanilla SCSS code changes, it should contain the following changes to make sure it's ready for the release:

  - `Feature 🎁`, `Breaking Change 💣`, `Bug 🐛`, `Documentation 📝`, `Maintenance 🔨`.
  - if CSS class names are not changed it can be bugfix relesase (x.x.**X**)
  - if CSS class names are changed/added/removed it should be minor version (x.**X**.0)
  - see the [wiki for more details](https://github.com/canonical/vanilla-framework/wiki/Release-process#pre-release-tasks)


## Screenshots

<img width="618" alt="image" src="https://github.com/canonical/vanilla-framework/assets/83575/6ce39d36-7537-4661-a78a-ece5032c8e3c">

<img width="598" alt="image" src="https://github.com/canonical/vanilla-framework/assets/83575/05f6ed5b-3c97-458b-8643-ec9b89f02dd5">

<img width="340" alt="image" src="https://github.com/canonical/vanilla-framework/assets/83575/72e60366-6cf0-4efe-aca4-62376a735fcf">





[WD-8624]: https://warthogs.atlassian.net/browse/WD-8624?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ
